metals/src/main/scala/scala/meta/internal/metals/codeactions/ExtractMethodCodeAction.scala

```diff
@@ -0,0 +1,165 @@
+    trees: Trees
+) extends CodeAction {
+  override def kind: String = l.CodeActionKind.RefactorExtract
+  override def contribute(params: CodeActionParams, token: CancelToken)(implicit
+      ec: ExecutionContext
+  ): Future[Seq[l.CodeAction]] = Future {
+    val path = params.getTextDocument().getUri().toAbsolutePath
+    val range = params.getRange()
+
+    val toExtract: Option[List[Tree]] = {
+      if (range.getStart() == range.getEnd()) {
+        None
+      } else {
+        val tree: Option[Tree] = trees.get(path)
+        def loop(appl: Tree): Option[Tree] = {
+          appl.children.find(_.pos.encloses(range)) match {
+            case Some(child) =>
+              loop(child)
+            case None =>
+              Some(appl)
+          }
+        }
+        val enclosing = tree.flatMap(loop(_))
+        enclosing.map(_ match {
+          case Term.Block(stats) =>
+            stats.filter((s: Tree) => range.encloses(s.pos.toLSP))
+          case Template(_, _, _, stats) =>
+            stats.filter((s: Tree) => range.encloses(s.pos.toLSP))
+          case ap if returnsValue(ap) => List(ap)
+          case _ => Nil
+        })
+      }
+    }
+
+    val edits = {
+      for {
+        exprs <- toExtract
+        last <- exprs.lastOption
+        head <- exprs.headOption if returnsValue(last)
+        scopeOptions = enclosingList(head).flatMap(enclosingDef(_))
+      } yield {
+        (scopeOptions, last, head)
+      }
+    }
+    edits
+      .map { case (scopes: List[(Tree, Tree)], apply: Tree, head: Tree) =>
+        scopes.map { case (defn, block) =>
+          val defnRange =
+            stats(block).find(_.pos.end >= head.pos.end).getOrElse(defn)
+          val scopeName = defnTitle(defn)
+
+          val codeAction = new l.CodeAction(
+            ExtractMethodCodeAction.title(scopeName)
+          )
+          codeAction.setKind(l.CodeActionKind.RefactorExtract)
+
+          codeAction.setCommand(
+            ServerCommands.ExtractMethod.toLSP(
+              ServerCommands.ExtractMethodParams(
+                params.getTextDocument(),
+                new l.Range(
+                  head.pos.toLSP.getStart(),
+                  apply.pos.toLSP.getEnd(),
+                ),
+                defnRange.pos.toLSP,
+              )
+            )
+          )
+          codeAction
+        }
+      }
+      .getOrElse(Nil)
+
+  }
+
+  private def defnTitle(defn: Tree): String = {
+    defn match {
+      case vl: Defn.Val =>
+        vl.pats.head match {
+          case Pat.Var(name) => s"val `${name}`"
+          case _ => "val"
+        }
+      case vr: Defn.Var =>
+        vr.pats.head match {
+          case Pat.Var(name) => s"var `${name}`"
+          case _ => "var"
+        }
+      case df: Defn.Def => s"method `${df.name}`"
+      case cl: Defn.Class => s"class `${cl.name}`"
+      case en: Defn.Enum => s"enum ${en.name}`"
+      case ob: Defn.Object => s"object `${ob.name}`"
+      case _ => "block"
```

Comment:
Maybe if we have multiple blocks we should differentiate it with the first line?